### ISSUE TITLE
r2r speed test

### ISSUE THREAD

**Issue by dengls24** [thu-nics/R2R#31]
I encountered some issues during the r2r speed test. The results showed that it was even faster than running slm alone. It seems there might be a problem with the hardware.

(r2r) dengl@asic-u3:~/6-mslm/R2R$ python script/playground/speed_benchmark.py --test_r2r --router_path resource/default_router.pt
CUDA available with 2 GPUs
Starting Speed Benchmark...
Test prompts: 10
Max new tokens: 100
Temperature: 0.0
Top-p: 1.0

=== Benchmarking R2R Dynamic Mode ===
Using 2 GPUs for SGLang, with 2 for reference and 1 for quick.
Loading quick model DeepSeek-R1-Distill-Qwen-1.5B...
[W1213 11:35:26.865416661 socket.cpp:759] [c10d] The client socket cannot be initialized to connect to [localhost]:30787 (errno: 97 - Address family not supported by protocol).
Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]
Loading safetensors checkpoint shards: 100% Completed | 1/1 [00:00<00:00,  1.11it/s]
Loading safetensors checkpoint shards: 100% Completed | 1/1 [00:00<00:00,  1.11it/s]

Capturing batches (avail_mem=62.31 GB): 100%|████████████████████| 23/23 [00:09<00:00,  2.36it/s]
Capturing batches (avail_mem=61.22 GB): 100%|████████████████████| 23/23 [00:08<00:00,  2.78it/s]
Loading reference model DeepSeek-R1-Distill-Qwen-32B...
[W1213 11:35:59.095921115 socket.cpp:759] [c10d] The client socket cannot be initialized to connect to [localhost]:29500 (errno: 97 - Address family not supported by protocol).
[W1213 11:36:01.855120443 socket.cpp:759] [c10d] The client socket cannot be initialized to connect to [localhost]:29500 (errno: 97 - Address family not supported by protocol).
The memory capacity is unbalanced. Some GPUs may be occupied by other processes. min_per_gpu_memory=60.57098388671875, local_gpu_memory=72.6241455078125, local_gpu_memory * 0.9=65.36173095703126
Loading safetensors checkpoint shards:   0% Completed | 0/8 [00:00<?, ?it/s]
Loading safetensors checkpoint shards:  12% Completed | 1/8 [00:02<00:15,  2.21s/it]
Loading safetensors checkpoint shards:  25% Completed | 2/8 [00:04<00:11,  1.99s/it]
Loading safetensors checkpoint shards:  38% Completed | 3/8 [00:06<00:10,  2.18s/it]
Loading safetensors checkpoint shards:  50% Completed | 4/8 [00:08<00:07,  1.93s/it]
Loading safetensors checkpoint shards:  62% Completed | 5/8 [00:10<00:06,  2.05s/it]
Loading safetensors checkpoint shards:  75% Completed | 6/8 [00:11<00:03,  1.65s/it]
Loading safetensors checkpoint shards:  88% Completed | 7/8 [00:13<00:01,  1.81s/it]
Loading safetensors checkpoint shards: 100% Completed | 8/8 [00:15<00:00,  1.92s/it]
Loading safetensors checkpoint shards: 100% Completed | 8/8 [00:15<00:00,  1.93s/it]

Sliding Window Attention is enabled but not implemented for `sdpa`; unexpected results may be encountered.
Successfully copied weights from deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B embeddings
Added 'token_embeddings.weight' to state_dict
Using provided threshold: 0.9
Using input types: ['hidden_states', 'token', 'logits']
Testing prompt 1/10: What is the capital of France?...
Dynamic SGLang: Generating tokens: 100%|███████████████████████| 100/100 [00:02<00:00, 40.01it/s]
  Output tokens: 88, Time: 2.50s, Speed: 35.18 tok/s
Testing prompt 2/10: Explain the concept of machine learning in simple ...
Dynamic SGLang: Generating tokens: 100%|███████████████████████| 100/100 [00:01<00:00, 56.41it/s]
  Output tokens: 84, Time: 1.77s, Speed: 47.35 tok/s
Testing prompt 3/10: Write a short story about a robot learning to pain...
Dynamic SGLang: Generating tokens: 100%|███████████████████████| 100/100 [00:02<00:00, 42.45it/s]
  Output tokens: 84, Time: 2.36s, Speed: 35.64 tok/s
Testing prompt 4/10: How do neural networks work?...
Dynamic SGLang: Generating tokens: 100%|███████████████████████| 100/100 [00:01<00:00, 84.68it/s]
  Output tokens: 89, Time: 1.18s, Speed: 75.26 tok/s
Testing prompt 5/10: What are the main differences between Python and J...
Dynamic SGLang: Generating tokens: 100%|███████████████████████| 100/100 [00:01<00:00, 77.55it/s]
  Output tokens: 85, Time: 1.29s, Speed: 65.85 tok/s
Testing prompt 6/10: Describe the process of photosynthesis....
Dynamic SGLang: Generating tokens: 100%|███████████████████████| 100/100 [00:01<00:00, 73.49it/s]
  Output tokens: 88, Time: 1.36s, Speed: 64.62 tok/s
Testing prompt 7/10: What is the theory of relativity?...
Dynamic SGLang: Generating tokens: 100%|███████████████████████| 100/100 [00:01<00:00, 59.35it/s]
  Output tokens: 87, Time: 1.69s, Speed: 51.60 tok/s
Testing prompt 8/10: How does blockchain technology work?...
Dynamic SGLang: Generating tokens: 100%|███████████████████████| 100/100 [00:01<00:00, 85.38it/s]
  Output tokens: 89, Time: 1.17s, Speed: 75.91 tok/s
Testing prompt 9/10: Explain the concept of recursion in programming....
Dynamic SGLang: Generating tokens: 100%|██████████████████████| 100/100 [00:00<00:00, 110.13it/s]
  Output tokens: 86, Time: 0.91s, Speed: 94.58 tok/s
Testing prompt 10/10: What are the benefits of renewable energy?...
Dynamic SGLang: Generating tokens: 100%|███████████████████████| 100/100 [00:01<00:00, 61.89it/s]
  Output tokens: 87, Time: 1.62s, Speed: 53.81 tok/s

============================================================
BENCHMARK SUMMARY
============================================================

R2R:
  Total tokens: 867
  Total time: 15.85s
  Average speed: 59.98 tok/s
  Speed range: 35.18 - 94.58 tok/s

Detailed results saved to /u3disk/dengl/6-mslm/R2R/speed_benchmark_20251213_113656.json
[rank0]:[W1213 11:36:57.665554093 ProcessGroupNCCL.cpp:1496] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
nanobind: leaked 1 instances!
 - leaked instance 0x7e7ffc1c6898 of type "xgrammar.xgrammar_bindings.GrammarCompiler"
nanobind: leaked 1 types!
 - leaked type "xgrammar.xgrammar_bindings.GrammarCompiler"
nanobind: leaked 7 functions!
 - leaked function "__init__"
 - leaked function "compile_builtin_json_grammar"
 - leaked function "compile_grammar"
 - leaked function "compile_structural_tag"
 - leaked function "compile_regex"
 - leaked function "clear_cache"
 - leaked function "compile_json_schema"
nanobind: this is likely caused by a reference counting issue in the binding code.
/u3disk/dengl/.conda/envs/r2r/lib/python3.10/multiprocessing/resource_tracker.py:224: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
/u3disk/dengl/.conda/envs/r2r/lib/python3.10/multiprocessing/resource_tracker.py:224: UserWarning: resource_tracker: There appear to be 1 leaked shared_memory objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '

**Comment by ggeeyyi** [comment]
Thanks for your report! but I’m also a bit concerned about the experimental settings. Did you ensure that both R2R and the SLM-only baseline use the same `mem_fraction_static`, and that both are **properly warmed up** for a fair comparison? Also, I think this is a reasonable concern because in our inference setup, we actually set `disable_radix_cache` and `disable_cuda_graph` to **false** to enable inference acceleration.
